nuto/mechanics/tools/QuasistaticSolver.cpp

```diff
@@ -44,134 +48,104 @@ void QuasistaticSolver::SetGlobalTime(double globalTime)
 
-    auto hessian0 = mProblem.Hessian0(v, mDofs, globalTime, timeStep);
-    auto hJJ = ToEigen(hessian0.JJ, mDofs);
-    auto hJK = ToEigen(hessian0.JK, mDofs);
-    auto hKJ = ToEigen(hessian0.KJ, mDofs);
-    auto hKK = ToEigen(hessian0.KK, mDofs);
-    auto cMat = ToEigen(mCmat, mDofs);
-    Eigen::SparseMatrix<double> hessianMod = hJJ - cMat.transpose() * hKJ - hJK * cMat + cMat.transpose() * hKK * cMat;
+    // update time step
+    double newTimeStep = newGlobalTime - mGlobalTime;
 
-    DofVector<double> deltaBrhs;
-    for (auto dof : mDofs)
-        deltaBrhs[dof] = mConstraints.GetRhs(dof, globalTime + timeStep) - mConstraints.GetRhs(dof, globalTime);
+    // compute residual for new time step (in particular, these are changing external forces)
+    auto gradient = mProblem.Gradient(mX, mDofs, newGlobalTime, newTimeStep);
 
-    Eigen::VectorXd residualConstrained = -(hJK - cMat.transpose() * hKK) * ToEigen(deltaBrhs, mDofs);
+    DofVector<double> trialU = mX - solver.SolveTrialState(hessian0, gradient, mGlobalTime, newGlobalTime);
 
-    return std::make_pair(hessianMod, residualConstrained);
+    return trialU;
 }
 
-Eigen::VectorXd QuasistaticSolver::Residual(const Eigen::VectorXd& x)
+DofVector<double> QuasistaticSolver::Residual(const DofVector<double>& u)
 {
-    auto gradient = mProblem.Gradient(ToGlobalDofVector(x), mDofs, mGlobalTime + mTimeStep, mTimeStep);
-    return ToEigen(gradient.J, mDofs) - ToEigen(mCmat, mDofs).transpose() * ToEigen(gradient.K, mDofs);
+    return mProblem.Gradient(u, mDofs, mGlobalTime + mTimeStep, mTimeStep);
 }
 
 
-Eigen::SparseMatrix<double> QuasistaticSolver::Derivative(const Eigen::VectorXd& x)
+DofMatrixSparse<double> QuasistaticSolver::Derivative(const DofVector<double>& u)
 {
-    auto hessian0 = mProblem.Hessian0(ToGlobalDofVector(x), mDofs, mGlobalTime + mTimeStep, mTimeStep);
-    auto hJJ = ToEigen(hessian0.JJ, mDofs);
-    auto hJK = ToEigen(hessian0.JK, mDofs);
-    auto hKJ = ToEigen(hessian0.KJ, mDofs);
-    auto hKK = ToEigen(hessian0.KK, mDofs);
-    auto cMat = ToEigen(mCmat, mDofs);
-    return hJJ - cMat.transpose() * hKJ - hJK * cMat + cMat.transpose() * hKK * cMat;
+    return mProblem.Hessian0(u, mDofs, mGlobalTime + mTimeStep, mTimeStep);
 }
 
-void QuasistaticSolver::UpdateHistory(const Eigen::VectorXd& x)
+void QuasistaticSolver::UpdateHistory(const DofVector<double>& x)
 {
-    mProblem.UpdateHistory(ToGlobalDofVector(x), mDofs, mGlobalTime + mTimeStep, mTimeStep);
+    mProblem.UpdateHistory(x, mDofs, mGlobalTime + mTimeStep, mTimeStep);
 }
 
-double QuasistaticSolver::Norm(const Eigen::VectorXd& residual) const
+double QuasistaticSolver::Norm(const DofVector<double>& residual) const
 {
-    return residual.cwiseAbs().maxCoeff();
+    // all these conversions should be removed at some point, workaround to remove JK
+    Eigen::VectorXd tmp = ToEigen(residual, mDofs);
+    auto C = ToEigen(mCmatUnit, mDofs);
+    return (C.transpose() * tmp).norm();
```

Comment:
I like that now pass `Dof` types instead of `Eigen`. We could now use this feature to introduce separate `mTolerance` for each dof type. Not urgent though.